File: java/src/main/java/vsdk/toolkit/raycasting/photonMap/PhotonKDTree.java
/**
Photon kd-tree : specialized kd-tree with some photon map specific additions
*/

package vsdk.toolkit.raycasting.photonMap;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

import vsdk.toolkit.common.dataStructures.KDTree;
import vsdk.toolkit.common.linealAlgebra.Vector3D;

public class PhotonKDTree extends KDTree {
    private static class NodeData {
        Photon photon;
        short flags;
        float[] point;

        NodeData(Photon photon, short flags) {
            this.photon = photon;
            this.flags = flags;
            Vector3D p = photon.pos();
            this.point = new float[] {p.x, p.y, p.z};
        }
    }

    private static class QueryCandidate {
        Photon photon;
        float distance;

        QueryCandidate(Photon photon, float distance) {
            this.photon = photon;
            this.distance = distance;
        }
    }

    private static NormalQuery qdat_s = new NormalQuery();
    private final List<NodeData> nodes;

    // Distance calculation COPY FROM kdtree.C !
    private static float sqrDistance3D(float[] a, float[] b) {
        float result;
        float tmp;

        tmp = a[0] - b[0];
        result = tmp * tmp;

        tmp = a[1] - b[1];
        result += tmp * tmp;

        tmp = a[2] - b[2];
        result += tmp * tmp;

        return result;
    }

    public PhotonKDTree(int dataSize, boolean copyData) {
        super(dataSize, copyData);
        nodes = new ArrayList<>();
    }

    @Override
    public void addPoint(Object data, short flags) {
        if ( !(data instanceof Photon) ) {
            throw new IllegalArgumentException("PhotonKDTree only accepts Photon data");
        }
        nodes.add(new NodeData((Photon)data, flags));
    }

    @Override
    public int query(float[] point, int n, Object[] results) {
        return query(point, n, results, null, KD_MAX_RADIUS, (short)0);
    }

    @Override
    public int query(float[] point, int n, Object[] results, float[] inDistances, float radius, short excludeFlags) {
        List<QueryCandidate> candidates = new ArrayList<>();

        for ( NodeData node : nodes ) {
            if ( (node.flags & excludeFlags) != 0 ) {
                continue;
            }
            float dist = sqrDistance3D(node.point, point);
            if ( dist < radius ) {
                candidates.add(new QueryCandidate(node.photon, dist));
            }
        }

        candidates.sort(Comparator.comparingDouble(c -> c.distance));

        int found = Math.min(n, candidates.size());
        if ( found <= 0 ) {
            return 0;
        }

        float[] usedDistances = inDistances;
        if ( usedDistances == null ) {
            usedDistances = new float[Math.max(1, found)];
        }

        // C kd-tree query keeps the farthest of the selected photons at index 0.
        QueryCandidate farthest = candidates.get(found - 1);
        results[0] = farthest.photon;
        usedDistances[0] = farthest.distance;

        int outIndex = 1;
        for ( int i = 0; i < found - 1; i++ ) {
            results[outIndex] = candidates.get(i).photon;
            usedDistances[outIndex] = candidates.get(i).distance;
            outIndex++;
        }

        return found;
    }

    @Override
    public void iterateNodes(NodeCallback callback, Object data) {
        for ( NodeData node : nodes ) {
            callback.call(data, node.photon);
        }
    }

    @Override
    public void balance() {
        // This Java migration keeps a linear container. No explicit balancing needed.
    }

    /**
Find the nearest photon with a similar normal constraint
returns nullptr is no appropriate photon was found (should barely ever happen)
*/
    public IrrPhoton
    normalPhotonQuery(
        Vector3D position,
        Vector3D normal,
        float threshold,
        float maxR2)
    {
        // Fill qdat_s
        qdat_s.photon = null;
        qdat_s.normal = new Vector3D(normal.x, normal.y, normal.z);
        qdat_s.point = new float[] {position.x, position.y, position.z};
        qdat_s.threshold = threshold;
        qdat_s.maximumDistance = maxR2;

        for ( NodeData node : nodes ) {
            if ( !(node.photon instanceof IrrPhoton) ) {
                continue;
            }

            float dist = sqrDistance3D(node.point, qdat_s.point);

            // Normal constraint
            IrrPhoton photon = (IrrPhoton)node.photon;
            if ( dist < qdat_s.maximumDistance &&
                 (photon.Normal().dotProduct(qdat_s.normal) > qdat_s.threshold ) ) {
                // Replace point if distance < maxdist AND normal is similar
                qdat_s.maximumDistance = dist;
                qdat_s.photon = photon;
            }
        }

        return qdat_s.photon;
    }
}

File: typescript/base/src/main/vsdk/toolkit/raycasting/photonMap/PhotonKDTree.ts
/**
Photon kd-tree : specialized kd-tree with some photon map specific additions
*/

import { KDTree, NodeCallback } from "../../common/dataStructures/KDTree";
import { Vector3D } from "../../common/linealAlgebra/Vector3D";
import { IrrPhoton } from "./IrrPhoton";
import { NormalQuery } from "./NormalQuery";
import { Photon } from "./Photon";

class NodeData {
  public photon: Photon;
  public flags: number;
  public point: number[];

  public constructor(photon: Photon, flags: number) {
    this.photon = photon;
    this.flags = flags;
    const p = photon.pos();
    this.point = [p.x, p.y, p.z];
  }
}

class QueryCandidate {
  public photon: Photon;
  public distance: number;

  public constructor(photon: Photon, distance: number) {
    this.photon = photon;
    this.distance = distance;
  }
}

export class PhotonKDTree extends KDTree {
  private static qdat_s = new NormalQuery();
  private readonly nodes: NodeData[];

  private static sqrDistance3DPhoton(a: number[], b: number[]): number {
    const a0 = a[0] ?? 0.0;
    const b0 = b[0] ?? 0.0;
    let tmp = a0 - b0;
    let result = tmp * tmp;

    const a1 = a[1] ?? 0.0;
    const b1 = b[1] ?? 0.0;
    tmp = a1 - b1;
    result += tmp * tmp;

    const a2 = a[2] ?? 0.0;
    const b2 = b[2] ?? 0.0;
    tmp = a2 - b2;
    result += tmp * tmp;

    return result;
  }

  public constructor(dataSize: number, copyData: boolean) {
    super(dataSize, copyData);
    this.nodes = [];
  }

  public override addPoint(data: unknown, flags = 0): void {
    if (!(data instanceof Photon)) {
      throw new globalThis.Error("PhotonKDTree only accepts Photon data");
    }
    this.nodes.push(new NodeData(data, flags));
  }

  public override query(point: number[], n: number, results: unknown[]): number;
  public override query(
    point: number[],
    n: number,
    results: unknown[],
    inDistances: number[] | null,
    radius: number,
    excludeFlags: number
  ): number;
  public override query(
    point: number[],
    n: number,
    results: unknown[],
    inDistances: number[] | null = null,
    radius = KDTree.KD_MAX_RADIUS,
    excludeFlags = 0
  ): number {
    const candidates: QueryCandidate[] = [];

    for (let i = 0; i < this.nodes.length; i++) {
      const node = this.nodes[i];
      if (node === undefined) {
        continue;
      }
      if ((node.flags & excludeFlags) !== 0) {
        continue;
      }
      const dist = PhotonKDTree.sqrDistance3DPhoton(node.point, point);
      if (dist < radius) {
        candidates.push(new QueryCandidate(node.photon, dist));
      }
    }

    candidates.sort((a, b) => a.distance - b.distance);

    const found = globalThis.Math.min(n, candidates.length);
    if (found <= 0) {
      return 0;
    }

    const usedDistances = inDistances === null ? new Array<number>(globalThis.Math.max(1, found)).fill(0.0) : inDistances;

    const farthest = candidates[found - 1];
    if (farthest === undefined) {
      return 0;
    }
    results[0] = farthest.photon;
    usedDistances[0] = farthest.distance;

    let outIndex = 1;
    for (let i = 0; i < found - 1; i++) {
      const candidate = candidates[i];
      if (candidate === undefined) {
        continue;
      }
      results[outIndex] = candidate.photon;
      usedDistances[outIndex] = candidate.distance;
      outIndex++;
    }

    return found;
  }

  public override iterateNodes(callback: NodeCallback, data: unknown): void {
    for (let i = 0; i < this.nodes.length; i++) {
      const node = this.nodes[i];
      if (node !== undefined) {
        callback.call(data, node.photon);
      }
    }
  }

  public override balance(): void {
    // This Java migration keeps a linear container. No explicit balancing needed.
  }

  public normalPhotonQuery(
    position: Vector3D,
    normal: Vector3D,
    threshold: number,
    maxR2: number
  ): IrrPhoton | null {
    PhotonKDTree.qdat_s.photon = null;
    PhotonKDTree.qdat_s.normal = new Vector3D(normal.x, normal.y, normal.z);
    PhotonKDTree.qdat_s.point = [position.x, position.y, position.z];
    PhotonKDTree.qdat_s.threshold = threshold;
    PhotonKDTree.qdat_s.maximumDistance = maxR2;

    for (let i = 0; i < this.nodes.length; i++) {
      const node = this.nodes[i];
      if (node === undefined) {
        continue;
      }
      if (!(node.photon instanceof IrrPhoton)) {
        continue;
      }

      const dist = PhotonKDTree.sqrDistance3DPhoton(node.point, PhotonKDTree.qdat_s.point as number[]);
      const photon = node.photon;
      if (dist < PhotonKDTree.qdat_s.maximumDistance
        && photon.Normal().dotProduct(PhotonKDTree.qdat_s.normal) > PhotonKDTree.qdat_s.threshold) {
        PhotonKDTree.qdat_s.maximumDistance = dist;
        PhotonKDTree.qdat_s.photon = photon;
      }
    }

    return PhotonKDTree.qdat_s.photon;
  }

  // Java parity helper name; instance scope avoids static private-name clash with KDTree.
  private sqrDistance3D(a: number[], b: number[]): number {
    return PhotonKDTree.sqrDistance3DPhoton(a, b);
  }
}
